[READY] Raise warnings on FileReadyToParse for completers not supporting diagnostics

puremourning opened the pull request:
Fixes #1829 

## Background

Detailed analysis is available on [the issue](https://github.com/Valloric/YouCompleteMe/issues/1829). But principally: if the filetype being edited is not a filetype suporting YCM native diagnostics, then warnings, errors, etc. (including the "unknown extra conf" warning/prompt) are not displayed after the "FileReadyToParse" `event_notification`

## Resolution

The approach is: for filetypes which do not support diagnostics, rather than ignoring all `FileReadyToParse` event responses, simply call the `Response` method, which raises the appropriate error or handles the "confirm extra-conf" behaviour. Any other data returned by the completer (such as diagnostics!), are simply ignored.

## Manual Testing

Tested with tern completer which raises an error if the `.tern-project` file is not found.
Tested with a modified tern completer with the following `OnFileReadyToParse`:

```python
  def OnFileReadyToParse( self, request_data ):
    self._WarnIfMissingTernProject()

    module = extra_conf_store.ModuleForSourceFile( request_data[ 'filepath' ] )
    if not module:
      if not self._raised_extra_conf_warning:
        self._raised_extra_conf_warning = True
        raise NoExtraConfDetected
      return

    # use module!
```

Results all as expected:

- with no `.tern-config`: tern config warning (once)
- with `.tern-config`, but no `.ycm_extra_conf`: no extra conf warning (once)
- with `.tern-config` and `.ycm_extra_conf`: confirmation prompt (once)

## Automated Testing

This was a marathon. Or a mockathon, I suppose.

Briefly:

- I wrote a `event_notification_test.py` which tests these new `YouCompleteMe` object behaviours, via a complex series of mocking
- I had to refactor the `postcomplete_tests.py` to prevent it from leaving `MagicMock` instances all over the place because this lead to the above test failing (apparently) randomly
- In order to debug the above shenanigans I updated `run_tests.py` to allow passing arguments to nosetests, like we do in `ycmd`

There are more comments in the code and on the various commit messages, explaining what's going on.

With regard to style, there is quite a lot of:

```python

@contextlib.contextmanager
def something():
  with patch(this):
    with patch(that):
      with patch(the_other):
        yield

@patch( foo )
def do_a_test( foo ):
  with something():
    do_a_thing() 
    assert foo.called()
```

The reason for this relative ugliness is that there is a lot of boilerplate mocking code. Because of a number of foibles of `mock` and the versions of python we support, I found this to be the least hacky way of ensuring that everything was cleaned up and was relatively sane to read and maintain.

Explanation for the `with`-ception that is going on:

- @patch is my preferred method, but it is more difficult to comment well and doesn't work when used with `@contextlib.contextmanager`, so you have to use `with`
- multiple managers in a single `with` statement is not supported in Python 2.6
- `contextlib.nested` is deprecated and buggy (apparently)

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/valloric/youcompleteme/1848)
<!-- Reviewable:end -->

Valloric commented:
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[python/ycm/tests/event_notification_test.py, line 19 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1848#-K6H1mIJeIs6PfUtqbEv:-K6H1mIJeIs6PfUtqbEw:1132071504) ([raw file](https://github.com/valloric/youcompleteme/blob/3b5f1c0431507b8edc5511bd0b24f2effa72cbd4/python/ycm/tests/event_notification_test.py#L19)):*
AFAIR we try to avoid relative imports since they're not considered a Python best practice (they have a tendency to blow up).

---

*[python/ycm/tests/event_notification_test.py, line 28 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1848#-K6H2Kt3JdOM1l1jMAFP:-K6H2Kt3JdOM1l1jMAFQ:1579994090) ([raw file](https://github.com/valloric/youcompleteme/blob/3b5f1c0431507b8edc5511bd0b24f2effa72cbd4/python/ycm/tests/event_notification_test.py#L28)):*
Why the parens?

---

*[python/ycm/youcompleteme.py, line 495 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1848#-K6H37jfsDZWaCGgaDHD:-K6H37jfsDZWaCGgaDHE:-1690954998) ([raw file](https://github.com/valloric/youcompleteme/blob/3b5f1c0431507b8edc5511bd0b24f2effa72cbd4/python/ycm/youcompleteme.py#L495)):*
My memory is vague, but if I recall correctly, I started off by not having the client know which filetypes have semantic completion/diagnostics. But it became necessary to have that info in the client for the sake of Vim performance. Can't remember the details.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1848)*
<!-- Sent from Reviewable.io -->

Valloric commented:
LGTM with minor comments. I'd like someone else from the team to take a look as well though. @micbou, @vheon ?

---

Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1848#-:-K6H4HJ96glA1MlHtxPX:1094374387)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Thanks. I've updated.

---

Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[python/ycm/tests/event_notification_test.py, line 19 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1848#-K6H1mIJeIs6PfUtqbEv:-K6Jb1zS6M_RQZeXRQ-K:-691306461) ([raw file](https://github.com/valloric/youcompleteme/blob/3b5f1c0431507b8edc5511bd0b24f2effa72cbd4/python/ycm/tests/event_notification_test.py#L19)):*
Quite right. Not sure where I got this from ... Fixed.

---

*[python/ycm/tests/event_notification_test.py, line 28 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1848#-K6H2Kt3JdOM1l1jMAFP:-K6Jb5s8-iNDLfYiGQ-y:1762298244) ([raw file](https://github.com/valloric/youcompleteme/blob/3b5f1c0431507b8edc5511bd0b24f2effa72cbd4/python/ycm/tests/event_notification_test.py#L28)):*
Good question. Removed.

---

*[python/ycm/youcompleteme.py, line 495 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1848#-K6H37jfsDZWaCGgaDHD:-K6JbmxFFYFUJrkIYatz:-557964412) ([raw file](https://github.com/valloric/youcompleteme/blob/3b5f1c0431507b8edc5511bd0b24f2effa72cbd4/python/ycm/youcompleteme.py#L495)):*
I've removed the FIXME comment and tried to better describe what is currently going on. On reflection, the FIXME doesn't actually belong here anyway, and isn't really all that helpful in isolation.

For reference, here are some notes I wrote down:

Currently:
 - YCM queries the server a given filetype to determine if it has a semantic completer via `/semantic_completion_available` (called from `YouCompleteMe.FiletypeCompleterExistsForFiletype`)
- YCM has hard-coded list of filetypes which support diagnostics, from `autoload/youcompleteme.py`:

```vim
let s:diagnostic_ui_filetypes = {
      \ 'cpp': 1,
      \ 'cs': 1,
      \ 'c': 1,
      \ 'objc': 1,
      \ 'objcpp': 1,
      \ }
```

My thoughts were that we could use `/semantic_completion_available` to return whether or no diagnostics are supported too, but the comments in the `OnFileReadyToParse` seem to imply there there is some sort of implicit circular dependency (which TBH I haven't quite fathomed):

```vim
  " Order is important here; we need to extract any done diagnostics before
  " reparsing the file again. If we sent the new parse request first, then
  " the response would always be pending when we called
  " UpdateDiagnosticNotifications.
  call s:UpdateDiagnosticNotifications() " <-- benj: Uses the above mapping

  let buffer_changed = b:changedtick != b:ycm_changedtick.file_ready_to_parse
  if buffer_changed
    py ycm_state.OnFileReadyToParse() " benj: <-- would/might post /semantic_completion_available
  endif
```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1848#-:-K6JdxPpHDCi8B39CLXr:-1359318266)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Reviewed 2 of 5 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1848)*
<!-- Sent from Reviewable.io -->

vheon commented:
I've found just a typo in a comment. Other than that <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 2 of 5 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[python/ycm/tests/event_notification_test.py, line 49 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1848#-K6K2QGgF_synIXE0-Rj:-K6K2QGhCt1SqzlirQ45:255056475) ([raw file](https://github.com/valloric/youcompleteme/blob/2063ae8b9122b3879ccb412d8d0602c723ab7c44/python/ycm/tests/event_notification_test.py#L49)):*
typo `wby`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1848#-:-K6K3lTI75eBSLq3ZkDa:1647928894)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Typo fixed. I think this is now ready to merge.

---

Review status: 4 of 5 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1848#-:-K6KoIL9bYvL64CkLlje:-1344982804)*
<!-- Sent from Reviewable.io -->

vheon commented:
@Valloric already gave his <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> so I say les't call @homu r+

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1848#-:-K6KutYPaeHkE9EMu3Rt:-1723994313)*
<!-- Sent from Reviewable.io -->

homu commented:
:pushpin: Commit 0b07b74 has been approved by `vheon`

<!-- @homu r=vheon 0b07b7467d65d798f5ee3c19a3170432b874701d -->

homu commented:
:zap: Test exempted - [status](https://travis-ci.org/Valloric/YouCompleteMe/builds/98721400)
